operations/research-and-development/product/development-process/secure-software-development.md

```diff
@@ -0,0 +1,90 @@
+
+In practice, defining build-time dependencies with an explicit hash is the preferred solution. When specifying hashes to existing dependencies that did not use hash-based references before, picking the hash of the currently used dependency is a good starting point as it is often not practical to do a full audit of all dependencies. Picking a hash and sticking to it doesn't ensure that the hash represents a safe version, but it does ensures that a possibly malicious later version is not introduced in the build process afterwards.
+
+Many dependency managers support automatic dependency pinning so that a developer doesn't need to explicitly find and set dependency hashes. For example both Go and npm do, as described below.
+
+### Npm dependencies
+
+With npm, always use `npm ci` instead of `npm install` in build scripts and automated pipelines. This ensures that only packages matching the existing `package-lock.json` file are installed. See [the documentation on `npm ci`](https://docs.npmjs.com/cli/v8/commands/npm-ci) for more details.
+
+When adding new dependencies, prefer [exact versions](https://docs.npmjs.com/cli/v8/commands/npm-install#save-exact) over the default semver range.
+
+### Docker
+
+Docker image dependencies exist in Dockerfiles and elsewhere, such as in CircleCI config files. The [Dockpin](https://github.com/Jille/dockpin) tool can be used to pin Docker base images to their currently latest versions. To pin a dockerfile, run `dockpin docker pin -f Dockerfile`. To pin Docker base image dependencies elsewhere, run `dockpin docker resolve [base-image]`; for example `dockpin docker resolve ubuntu@20.04`.
+
+Here's a handy script to do this on CircleCI yaml files, replacing the image reference with one with a hash, and additionally commenting with the timestamp the hash was obtained:
+
+```
+#!/bin/bash
+
+ymlfile=$1
+
+if [ -z "$1" ] ; then
+    echo usage: $0 circleci-config-ymlfile
+    echo example: $0 ./.circleci/config.yml
+    exit 0
+fi
+
+for x in `grep -A1 docker $1|grep -- '- image:'|sed 's/.*image:\ \(.*\)/\1/g'|sed "s/'//g"|sort|uniq`;do
+    hashed=`dockpin docker resolve $x`
+    echo base $x hashed $hashed
+    sed -i "s~$x~$hashed\ #\ $x,\ `date`~g" $ymlfile
+done
+```
+
+### Go
+
+When adding a Go dependency using `go get`, the dependency can be installed specifying either a specific version (`go get example.com/theirmodule@v1.3.4`), the latest available version (`go get example.com/theirmodule@latest`), or a commit hash (`go get example.com/theirmodule@4cf76c2`) or branchname (`@branchname`). In all cases, the installed dependencies are added to the automatically generated `go.mod` file and the content hashes of the dependencies are written to the `go.sum` file. Make sure to commit both the `go.mod` and the `go.sum` file to your repository. This ensures that subsequent builds download exactly the same dependencies.
```

Comment:
Do we have any specific recommendation what approach would be recommended?

Reply:
i don't think this document can suggest a single best practice that applies in all cases

Reply:
That makes sense, we promote pinning in most examples for good reasons, shouldn't we discourage then using `latest` or is there a reason to use it with Go?

Reply:
Using `@latest` doesn't actually store a reference to `@latest` in `go.mod` or `go.sum` -- instead it resolves to whatever is the current latest version. Using it should be fine since the exact version still gets pinned as it would with any other version query.

Reply:
Perfect, thank you @jupenur!